pkg/registry/apis/provisioning/repository/github/github.go

```diff
@@ -0,0 +1,353 @@
+			return nil, repository.ErrFileNotFound
+		}
+
+		return nil, fmt.Errorf("get commits: %w", err)
+	}
+
+	ret := make([]provisioning.HistoryItem, 0, len(commits))
+	for _, commit := range commits {
+		authors := make([]provisioning.Author, 0)
+		if commit.Author != nil {
+			authors = append(authors, provisioning.Author{
+				Name:      commit.Author.Name,
+				Username:  commit.Author.Username,
+				AvatarURL: commit.Author.AvatarURL,
+			})
+		}
+
+		if commit.Committer != nil && commit.Author != nil && commit.Author.Name != commit.Committer.Name {
+			authors = append(authors, provisioning.Author{
+				Name:      commit.Committer.Name,
+				Username:  commit.Committer.Username,
+				AvatarURL: commit.Committer.AvatarURL,
+			})
+		}
+
+		ret = append(ret, provisioning.HistoryItem{
+			Ref:       commit.Ref,
+			Message:   commit.Message,
+			Authors:   authors,
+			CreatedAt: commit.CreatedAt.UnixMilli(),
+		})
+	}
+
+	return ret, nil
+}
+
+func (r *githubRepository) ensureBranchExists(ctx context.Context, branchName string) error {
+	if !git.IsValidGitBranchName(branchName) {
+		return &apierrors.StatusError{
+			ErrStatus: metav1.Status{
+				Code:    http.StatusBadRequest,
+				Message: "invalid branch name",
+			},
+		}
+	}
+
+	ok, err := r.gh.BranchExists(ctx, r.owner, r.repo, branchName)
+	if err != nil {
+		return fmt.Errorf("check branch exists: %w", err)
+	}
+
+	if ok {
+		logging.FromContext(ctx).Info("branch already exists", "branch", branchName)
+
+		return nil
+	}
+
+	srcBranch := r.config.Spec.GitHub.Branch
+	if err := r.gh.CreateBranch(ctx, r.owner, r.repo, srcBranch, branchName); err != nil {
+		if errors.Is(err, ErrResourceAlreadyExists) {
+			return &apierrors.StatusError{
+				ErrStatus: metav1.Status{
+					Code:    http.StatusConflict,
+					Message: "branch already exists",
+				},
+			}
+		}
+
+		return fmt.Errorf("create branch: %w", err)
+	}
+
+	return nil
+}
+
+func (r *githubRepository) LatestRef(ctx context.Context) (string, error) {
+	return r.gitRepo.LatestRef(ctx)
+}
+
+func (r *githubRepository) CompareFiles(ctx context.Context, base, ref string) ([]repository.VersionedFileChange, error) {
+	return r.gitRepo.CompareFiles(ctx, base, ref)
+}
+
+// ResourceURLs implements RepositoryWithURLs.
+func (r *githubRepository) ResourceURLs(ctx context.Context, file *repository.FileInfo) (*provisioning.ResourceURLs, error) {
+	cfg := r.config.Spec.GitHub
+	if file.Path == "" || cfg == nil {
+		return nil, nil
+	}
+
+	ref := file.Ref
+	if ref == "" {
+		ref = cfg.Branch
+	}
+
+	urls := &provisioning.ResourceURLs{
+		RepositoryURL: cfg.URL,
+		SourceURL:     fmt.Sprintf("%s/blob/%s/%s", cfg.URL, ref, file.Path),
+	}
+
+	if ref != cfg.Branch {
+		urls.CompareURL = fmt.Sprintf("%s/compare/%s...%s", cfg.URL, cfg.Branch, ref)
+
+		// Create a new pull request
+		urls.NewPullRequestURL = fmt.Sprintf("%s?quick_pull=1&labels=grafana", urls.CompareURL)
+	}
+
+	return urls, nil
+}
+
+// TODO: we should not need to clone
```

Comment:
We could probably rename that to stager or something like that. We can probably remove Clone and push options most likely.